docs/tutorials/on-device-training.md

```diff
@@ -0,0 +1,1406 @@
+            extendedCacheDir = extendedCacheDir + "/" + dirs.get(index)
+      }
+   }
+
+   // copy file from asset to cache dir in the same dir structure.
+   private fun copyAssetToCacheDir(assetFileName : String, cacheFileName : String): String {
+      mkCacheDir(cacheFileName)
+      val f = java.io.File("$cacheDir/$cacheFileName")
+      if (!f.exists()) {
+            try {
+               val modelFile = assets.open(assetFileName)
+               val size: Int = modelFile.available()
+               val buffer = ByteArray(size)
+               modelFile.read(buffer)
+               modelFile.close()
+               val fos = java.io.FileOutputStream(f)
+               fos.write(buffer)
+               fos.close()
+            } catch (e: Exception) {
+               throw RuntimeException(e)
+            }
+      }
+
+
+      return f.path
+   }
+
+   private fun copyFileOrDir(path: String): String {
+      val assetManager = assets
+      try {
+            val assets: Array<String>? = assetManager.list(path)
+            if (assets!!.size == 0) {
+               // asset is a file
+               copyAssetToCacheDir(path, path)
+            } else {
+               // asset is a dir. loop over dir and copy all files or sub dirs to cache dir
+               for (i in assets.indices) {
+                  var p: String
+                  p = if (path == "") "" else "$path/"
+                  copyFileOrDir(p + assets[i])
+               }
+            }
+      } catch (ex: java.io.IOException) {
+            Log.e("ortpersonalize", "I/O Exception", ex)
+      }
+
+      return "$cacheDir/$path"
+   }
+
+   external fun createSession(checkpointPath:String, trainModelPath: String, evalModelPath: String,
+                              optimizerModelPath: String, cacheDirPath: String): Long
+
+   external fun releaseSession(session:Long)
+
+   external fun performInference(session: Long, imageBuffer: FloatArray, batchSize: Int, channels: Int,
+                                 frameCols: Int, frameRows: Int, classes: Array<String>): String
+
+   external fun performTraining(session: Long, batches: FloatArray, labels: IntArray, batchSize: Int,
+                                 channels: Int, frameRows: Int, frameCols: Int) : Float
+
+   companion object {
+      // Used to load the 'ortpersonalize' library on application startup.
+      init {
+            System.loadLibrary("ortpersonalize")
+      }
+   }
+   ```
+
+   q. One last thing. Add the following in the `AndroidManifest.xml` file to use the camera:
+
+   ```xml
+   <uses-permission android:name="android.permission.CAMERA" />
+   <uses-feature android:name="android.hardware.camera" />
+   ```
+
+7. Running all this on a device.
+
+   a. Let's connect our Android device to the machine and run the application on the device.
+
+   b. Launching the application on the device should look like this:
+
+   ![Prediction - Cow](../../images/on-device-training-application-landing-page.jpg)
```

Comment:
This image renders as a super large image, any way to specify the width in markdown?